### PR TITLE
ci: Add Windows workflow

### LINKED ISSUE #770: Setup Flank Windows job on GitHub Actions
Setup a job that runs the Flank test suite on Windows using GitHub actions.

Use windows-latest

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #770 

If the workflow completes successfully it works

### CODE REVIEW

File: .github/workflows/windows_workflow.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-executable: "./test_runner/gradlew"
+          arguments: "-p test_runner check"
+
+      - name: Gradle clean build
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-executable: "./test_runner/gradlew"
+          arguments: "-p test_runner clean build"
```
Comment (piotradamczyk5) [inline .github/workflows/windows_workflow.yml:42]: gradle `build` run gradle `check` under the hood

`build` = `assemble` + `check`

File: .github/workflows/ubuntu-workflow.yml [hunk 2]
```diff
@@ -28,6 +28,12 @@ jobs:
           key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
           restore-keys: |
             ${{ runner.os }}-gradle-
+            
+      - name: Gradle check
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-executable: "./test_runner/gradlew"
+          arguments: "-p test_runner check"
```
Comment (piotradamczyk5) [inline .github/workflows/ubuntu-workflow.yml:36]: the same comment about `build` and `check`

you could check it using `--dry-run`

File: .github/workflows/windows_workflow.yml (hunk 1 above)
Comment (Sloox) [reply .github/workflows/windows_workflow.yml:42]: ill remove the check then 👍

**Comment by Sloox** [comment]
There are alot of build errors on Windows. Will need to fix this before we continue.

**Comment by adamfilipow92** [comment]
> There are alot of build errors on Windows. Will need to fix this before we continue.

I have configured Parallels with Windows so if you need any help tell me please

**Comment by Sloox** [comment]
> we only support Android on Windows. We don't ship the swift binaries for windows to run iOS tests.

it makes sense however it should be possible to have a action running a windows build for flank so its testable. But the amount to fix to get it working is expanding as its worked on.

**Comment by bootstraponline** [comment]
> it makes sense however it should be possible to have a action running a windows build for flank so its testable. But the amount to fix to get it working is expanding as its worked on.

agreed, I updated this to be a dragon 😂

File: test_runner/.gitignore [hunk 3]
```diff
@@ -8,3 +8,4 @@ src/test/kotlin/ftl/fixtures/error_result/*.xml
 src/test/kotlin/ftl/fixtures/success_result/*.xml
 *.json
 *-describe.adoc
+should_exists.txt
```
Comment (piotradamczyk5) [inline test_runner/.gitignore:11]: what is this ?

Comment (Sloox) [reply test_runner/.gitignore:11]: of all the weird quirks of windows. When you run test via gradle it adds this file. I just removed it via .gitignore.

Comment (pawelpasterz) [reply test_runner/.gitignore:11]: This should be removed at the end of some tests `UtilsTest`
```kotlin
companion object {
        @JvmStatic
        @AfterClass
        fun removeVerificationFile() {
            File(VERIFICATION_FILE).run {
                if (exists()) delete()
            }
        }
    }
```
Most likely test run terminated somewhere in the middle.
Just FYI

File: test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt [hunk 4]
```diff
@@ -93,7 +93,7 @@ internal class LogTableBuilderTest {
         val table = buildTable(*sampleColumns)
 
         // then
-        val tableLines = table.split(System.lineSeparator())
+        val tableLines = table.split("\n")
```
Comment (pawelpasterz) [inline test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt:96]: Isn't old implementation os agnostic? I probably miss something here, but why this change?

Comment (Sloox) [reply test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt:96]: Windows System.lineseperator = \n\r
All the tests cases make use of \n and quite alot have simply copy pasted text with \n insertions for new lines.
This is includes yaml files and other files that were originally edited on mac/linux. Therefore for almost ALL cases where System.lineseperator was used under windows. It failed to either split or compare etc. Hence why the simple change of \n fixes the majority of issues.

Comment (pawelpasterz) [reply test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt:96]: Thanks for explanation!

**Review by pawelpasterz — APPROVED** [review]

File: .github/workflows/windows_workflow.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          path: ~/.gradle/caches
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+
+      - name: Gradle clean build
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-executable: "./test_runner/gradlew"
+          arguments: "-p test_runner clean build"
```
Comment (piotradamczyk5) [inline .github/workflows/windows_workflow.yml:36]: Now we should use only arguments without executable
```suggestion
          arguments: ":test_runner:clean :test_runner:build"
```

File: test_runner/src/main/kotlin/ftl/args/ArgsHelper.kt [hunk 6]
```diff
@@ -73,10 +78,18 @@ object ArgsHelper {
     }
 
     fun evaluateFilePath(filePath: String): String {
-        var file = filePath.trim().replaceFirst(Regex("^~"), System.getProperty("user.home"))
+        val isWindows = System.getProperty("os.name").isWindows()
+
+        var file = if (isWindows) filePath.trim().convertToWindowsPath()
+        else
+            filePath.trim().replaceFirst(Regex("^~"), System.getProperty("user.home"))
         file = evaluateEnvVars(file)
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/ArgsHelper.kt:86]: could you wrap it to function?

File: test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt (hunk 4 above)
Comment (piotradamczyk5) [reply test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt:96]: @Sloox @pawelpasterz maybe we should create some variable with line end to always use it instead of `\n` or `System.lineSeparator()`?

File: .github/workflows/windows_workflow.yml (hunk 5 above)
Comment (piotradamczyk5) [reply .github/workflows/windows_workflow.yml:36]: please check ubuntu or macos for reference

File: test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt (hunk 4 above)
Comment (jan-gogo) [reply test_runner/src/test/kotlin/ftl/util/LogTableBuilderTest.kt:96]: You can, but IMO it's not really it is not necessary.

File: test_runner/src/test/kotlin/ftl/args/yml/ErrorParserTest.kt [hunk 7]
```diff
@@ -63,7 +64,7 @@ At line: 23, column: 3
               "version" : "test"
             }
         """.trimIndent()
-        Assert.assertEquals(exceptedMessage, actualMessage)
+        assert(equalsIgnoreNewlineStyle(exceptedMessage, actualMessage))
```
Comment (jan-gogo) [inline test_runner/src/test/kotlin/ftl/args/yml/ErrorParserTest.kt:67]: `assertEquals` is giving a more meaningful message. You have changed it to simple `assert` which not gives comprehensive information on failure. WDYF about changing function `equalsIgnoreNewlineStyle` to `assertEqualsIgnoreNewlineStyle`?

File: test_runner/gradle.properties [hunk 8]
```diff
@@ -9,7 +9,8 @@
 
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
-org.gradle.jvmargs=-Xmx2560m
+org.gradle.jvmargs=-Xmx2560m -Dfile.encoding=UTF-8
 kotlin.code.style=official
 org.gradle.parallel=true
-
+org.gradle.daemon=true
+org.gradle.caching=true
```
Comment (jan-gogo) [inline test_runner/gradle.properties:16]:
```suggestion
org.gradle.caching=true

```

File: test_runner/src/test/kotlin/ftl/args/yml/ErrorParserTest.kt (hunk 7 above)
Comment (Sloox) [reply test_runner/src/test/kotlin/ftl/args/yml/ErrorParserTest.kt:67]: Hmm Can do.

File: test_runner/gradle.properties (hunk 8 above)
Comment (Sloox) [reply test_runner/gradle.properties:16]: i swear i have this setting on for new lines. I think it might only apply to knwn files

File: test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt [hunk 9]
```diff
@@ -240,3 +241,14 @@ flank:
 }
 
 private fun validateYaml(args: IArgs.ICompanion, data: String): String = validateYaml(args, StringReader(data))
+
+fun assertEqualsIgnoreNewlineStyle(s1: String?, s2: String?) {
+    return Assert.assertEquals(equalsIgnoreNewlineStyle(s1, s2), true)
```
Comment (jan-gogo) [inline test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt:246]:
```suggestion
    Assert.assertNotNull(s1)
    Assert.assertNotNull(s2)
    Assert.assertEquals(normalizeLineEnds(s1), normalizeLineEnds(s2)
```

File: test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt [hunk 10]
```diff
@@ -240,3 +241,14 @@ flank:
...
+}
+fun equalsIgnoreNewlineStyle(s1: String?, s2: String?): Boolean {
+    return s1 != null && s2 != null && normalizeLineEnds(s1) == normalizeLineEnds(s2)
+}
```
Comment (jan-gogo) [inline test_runner/src/test/kotlin/ftl/doctor/DoctorTest.kt:250]:
```suggestion
```

**Comment by Sloox** [comment]
Actual integration tests running on Windows has moved to a new issue: #1134

**Review by piotradamczyk5 — APPROVED** [review]
